core/plugin_api.php

```diff
@@ -739,13 +755,19 @@ function plugin_upgrade( MantisPlugin $p_plugin ) {
 
 /**
  * Uninstall a plugin from the database.
+ * Return without action if given plugin is protected or not installed, unless
+ * it is a missing plugin, in which case we delete it regardless of protected
+ * status.
  * @param MantisPlugin $p_plugin Plugin basename.
+ *
  * @return void
  */
 function plugin_uninstall( MantisPlugin $p_plugin ) {
 	access_ensure_global_level( config_get_global( 'manage_plugin_threshold' ) );
 
-	if( !plugin_is_installed( $p_plugin->basename ) || plugin_protected( $p_plugin->basename ) ) {
+	if( !$p_plugin instanceof MissingClassPlugin &&
```

Comment:
I see how the different subclasses are being used. I'm not saying it's wrong.
I'm thinking, what about a different approach:
- where the validity status of the plugin is contained in the base MantisPlugin class
- those subclasses override that validity status
- The query for validity is used from the base class, don't need to know every subclass.

This would also allow a feature for a valid plugin, to manage it's own validity status, if it needs to perform specific test in order to work properly. Eg: valid configuration, additional resources, etc.

Reply:
Just to make sure if I understand properly, you want to have new method(s)s in the base MantisPlugin class, like *isValid()* to further encapsulate the logic within the class instead of doing it within plugin API ?

Reply:
it's not that i want it. It's an idea, just to discuss if it may make sense, or not worth it.

Reply:
Sorry, bad wording, I should not have said "want"... would like or mean would have been more appropriate.

Anyway, is my understanding of your idea correct or not ?

Reply:
> , is my understanding of your idea correct or not ?

yes 

> have new method(s)s in the base MantisPlugin class, like isValid() to further encapsulate the logic within the class instead of doing it within plugin API ?

I was thinking that the object provides both a "status" and an optional message, which more or less is what is being derived from each of the invalid plugin subclasses.

> isValid()

i dont kow if it's better a method like that, or just some class properties.

Reply:
Fixed.